### PR TITLE
Linux runtime code

### PR DESCRIPTION
**Pull request by rphh**

### CODE REVIEW

File: src/thread.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#include "thread.h"
+
+#include <iostream>
+
+namespace taste {
+void Thread::start(void (*method)()) {
+  m_method = nullptr;
```
Comment (Glamhoth) [inline src/thread.cpp:7]: Shouldn't this be `m_method = method`?

**Review by Glamhoth — CHANGES_REQUESTED** [review]
Additional comments:
1) This repo is missing `.clang-format`, you can copy one from our other repo (TASTE-Runtime-Common i.e.)
2) Docs are missing

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,59 @@
+cmake_minimum_required(VERSION 3.13 FATAL_ERROR)
+project(TASTE-Linux-Runtime VERSION 1.0.0 LANGUAGES C CXX)
+
+set(CMAKE_C_STANDARD 99)
```
[the remark this answers is not in the document]
Comment (rphh) [reply CMakeLists.txt:4]: Done

File: src/timer.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#ifndef TASTE_TIMER_H
+#define TASTE_TIMER_H
+
+#include <chrono>
+#include <thread>
+
+namespace taste {
+class Timer final {
+public:
+  Timer() = default;
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/timer.h]: True, removed.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+set(GCC_WARNINGS ${CLANG_WARNINGS}
+                 -Wmisleading-indentation
+                 -Wduplicated-cond
+                 -Wduplicated-branches
+                 -Wlogical-op)
+
+if(CMAKE_C_COMPILER_ID STREQUAL "Clang")
+    set(PROJECT_WARNINGS ${CLANG_WARNINGS})
+  elseif(CMAKE_C_COMPILER_ID STREQUAL "GNU")
```
[the remark this answers is not in the document]
Comment (rphh) [reply CMakeLists.txt]: fixed.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    set(PROJECT_WARNINGS ${GCC_WARNINGS})
+  endif()
+
+  target_compile_options(common_build_options INTERFACE ${PROJECT_WARNINGS})
```
[the remark this answers is not in the document]
Comment (rphh) [reply CMakeLists.txt]: fixed.

File: src/queue.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+  ~Queue() {}
+
+  Queue(const Queue &) = delete;
+  Queue(Queue &&) = delete;
+
+  Queue &operator=(const Queue &) = delete;
+  Queue &operator=(Queue &&) = delete;
+
+  void put(const Request<PARAMETER_SIZE> &request) {
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/queue.h]: Implementation outside class declaration is best what I can do.

File: src/queue.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    m_mutex.lock();
+    static Request<PARAMETER_SIZE> request;
+    memcpy(request.m_data, data, length);
+    request.m_length = length;
+
+    if (m_queue.size() > m_max_elements) {
+      std::cerr << "Message loss - queue is full, " << m_max_elements
+                << " allowed" << std::endl;
+      m_mutex.unlock();
+
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/queue.h]: CHanged.

File: src/timer.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    while (true) {
+      using namespace std::chrono_literals;
+      auto jitter = std::chrono::steady_clock::now() - start;
+      auto diff = interval - jitter;
+      if (diff > 0ms) {
+        std::this_thread::sleep_for(diff);
+      }
+      callback();
+
+      start = std::chrono::steady_clock::now();
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/timer.h]: Renamed

File: src/thread.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
+#ifndef TASTE_THREAD_H
+#define TASTE_THREAD_H
+
+#include <cstddef>
+#include <pthread.h>
+
+namespace taste {
+class Thread final {
+public:
+  Thread(int priority, size_t stack_size);
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/thread.h]: Done.

File: src/thread.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+void Thread::start(void (*method)(void *), void *param) {
+  m_method = method;
+  m_param = param;
+  create_thread(&Thread::method_wrapper_with_parameter,
+                reinterpret_cast<void *>(this));
+}
+
+Thread::Thread(int priority, size_t stack_size)
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/thread.cpp]: No. It is not necessary in TASTE. Thread shall be created, started and that's everything.

File: src/thread.cpp (hunk 1 above)
Comment (rphh) [reply src/thread.cpp]: No, comment added.

File: src/queue.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Glamhoth) [reply src/queue.h]: ack

File: src/Queue.h [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+ */
+template<size_t PARAMETER_SIZE>
+class Queue final {
+  public:
+    /**
+     * @brief Constructor
+     *
+     * @param max_elements    Maximum number of elements
+     */
+    Queue(size_t max_elements);
```
Comment (Glamhoth) [inline src/Queue.h:55]: const arg

File: src/Queue.h [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+    Queue(const Queue&) = delete;
```
Comment (Glamhoth) [inline src/Queue.h:57]: docs

File: src/Queue.h [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+    /**
+     * @brief Checks if queue is empty.
+     *
+     * @return true is queue is empty, otherwise false
+     */
+    bool is_empty() const;
+
+  private:
+    const size_t m_max_elements;
```
Comment (Glamhoth) [inline src/Queue.h:104]: docs

File: src/Lock.h [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TASTE_LOCK_H
+#define TASTE_LOCK_H
+
+/**
+ * @file     Lock.h
+ * @brief    Implemntation of lock required by TASTE function blocks.
```
Comment (Glamhoth) [inline src/Lock.h:28]: misspell

File: src/Lock.h [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    Lock();
+    Lock(const Lock&) = delete;
+    Lock(Lock&&) = delete;
+
+    Lock& operator=(const Lock&) = delete;
+    Lock& operator=(Lock&&) = delete;
+
+    /// @brief acquire lock
+    void lock();
+    /// @release lock
```
Comment (Glamhoth) [inline src/Lock.h:54]: @brief

File: src/Lock.h [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+ */
+class Lock final {
+  public:
+    /**
+     * @brief Default constructor.
+     *
+     * Constructed lock is in 'unlocked' state.
+     */
+    Lock();
+    Lock(const Lock&) = delete;
```
Comment (Glamhoth) [inline src/Lock.h:46]: docs

File: src/Lock.h [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    Lock& operator=(const Lock&) = delete;
+    Lock& operator=(Lock&&) = delete;
+
+    /// @brief acquire lock
+    void lock();
+    /// @release lock
+    void unlock();
+
+  private:
+    std::mutex m_mutex;
```
Comment (Glamhoth) [inline src/Lock.h:58]: docs

File: src/lock.h [hunk 18]
```diff
@@ -0,0 +1,25 @@
+#ifndef TASTE_LOCK_H
+#define TASTE_LOCK_H
+
+#include <mutex>
+
+namespace taste {
+class Lock final {
+public:
+  Lock();
+  ~Lock();
```
[the remark this answers is not in the document]
Comment (Glamhoth) [reply src/lock.h]: it's removed, ok

File: src/Timer.h [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+template<typename T>
+void
+Timer::run(std::chrono::milliseconds interval, T callback)
+{
+    auto currentTime = std::chrono::steady_clock::now();
+    while(true) {
+        using namespace std::chrono_literals;
+        auto jitter = std::chrono::steady_clock::now() - currentTime;
+        auto diff = interval - jitter;
```
Comment (Glamhoth) [inline src/Timer.h:59]: const

File: src/Thread.h [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+class Thread final {
+  public:
+    /**
+     * @brief Create a thread, without starting it
+     *
+     * @param priority     Priority of thread
+     * @param stack_size   Stack size for new thread in bytes
+     */
+    Thread(const int priority, const size_t stack_size);
+    Thread(const Thread&) = delete;
```
Comment (Glamhoth) [inline src/Thread.h:47]: docs

File: src/Thread.h [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * @brief Starts a thread with parameter
+     *
+     * @param method  A function which will be executed inside new thread.
+     * @param param   A parameter for function.
+     */
+    void start(void (*method)(void*), void* param);
+
+  private:
+    void create_thread(void* (*fn)(void*), void* param);
```
Comment (Glamhoth) [inline src/Thread.h:68]: do we document privates?

File: src/thread.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Glamhoth) [reply src/thread.cpp]: I meant that we should keep the same order of functions in both .h and .cpp files

File: src/thread.cpp (hunk 1 above)
Comment (Glamhoth) [reply src/thread.cpp]: ack

File: src/queue.h [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+        request = m_queue.front();
+        m_queue.pop();
+        lock.unlock();
+        return;
+      }
+    }
+  }
+
+  bool is_empty() const {
+    std::unique_lock<std::mutex> lock(m_mutex);
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/queue.h]: RAII

File: src/queue.h [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+
+  void get(Request<PARAMETER_SIZE> &request) {
+    std::unique_lock<std::mutex> lock(m_mutex);
+    while (true) {
+      if (m_queue.empty()) {
+        m_cv.wait(lock);
+      } else {
+        request = m_queue.front();
+        m_queue.pop();
+        lock.unlock();
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/queue.h]: I don't see an issue here:
`std::unique_lock<std::mutex> lock(m_mutex);` // locks m_mutex
`m_cv.wait(lock);` // unlocks the mutex, waits for notify, locks the mutex again, then returns

https://en.cppreference.com/w/cpp/thread/condition_variable

File: src/Timer.h (hunk 19 above)
Comment (rphh) [reply src/Timer.h]: done

File: src/Lock.h (hunk 17 above)
Comment (rphh) [reply src/Lock.h:59]: this is private

File: src/Lock.h (hunk 15 above)
Comment (rphh) [reply src/Lock.h]: fixed

File: src/Lock.h (hunk 14 above)
Comment (rphh) [reply src/Lock.h]: fixed

File: src/Queue.h (hunk 13 above)
Comment (rphh) [reply src/Queue.h:105]: this is private

File: src/Queue.h (hunk 11 above)
Comment (rphh) [reply src/Queue.h]: changed

File: src/thread.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (rphh) [reply src/thread.cpp]: Fixed

File: src/lock.h [hunk 24]
```diff
@@ -0,0 +1,25 @@
+#ifndef TASTE_LOCK_H
+#define TASTE_LOCK_H
+
+#include <mutex>
+
+namespace taste {
+class Lock final {
+public:
+  Lock();
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/lock.h]: changed

File: src/request.h [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+#define TASTE_REQUEST_H
+
+#include <cstdint>
+#include <cstring>
+
+namespace taste {
+template <size_t PARAMETER_SIZE> struct Request final {
+  Request() : m_length(0) {}
+
+  Request(const Request &other) : m_length(other.m_length) {
```
[the remark this answers is not in the document]
Comment (rphh) [reply src/request.h]: I've changed the array to std::array, now the constructors are defaulted,

File: src/Queue.h (hunk 12 above)
Comment (rphh) [reply src/Queue.h:59]: Done

File: src/Lock.h (hunk 16 above)
Comment (rphh) [reply src/Lock.h:49]: DOne

File: src/Thread.h (hunk 20 above)
Comment (rphh) [reply src/Thread.h:50]: Done

File: src/Thread.h (hunk 21 above)
Comment (rphh) [reply src/Thread.h:77]: No

File: src/queue.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (Glamhoth) [reply src/queue.h]: Ah, it's `unique_lock`, ok then

File: src/queue.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (Glamhoth) [reply src/queue.h]: You're right, I didn't notice that this is `unique_lock` and not our `Lock`

**Review by Glamhoth — APPROVED** [review]

File: src/Timer.h [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    template<typename T>
+    void run(const std::chrono::milliseconds interval, T callback);
+};
+
+template<typename T>
+void
+Timer::run(const std::chrono::milliseconds interval, T callback)
+{
+    auto currentTime = std::chrono::steady_clock::now();
```
Comment (Lurkerpas) [inline src/Timer.h:56]: I would call it reference time

File: src/Timer.h [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+    while(true) {
+        using namespace std::chrono_literals;
+        const auto jitter = std::chrono::steady_clock::now() - currentTime;
```
Comment (Lurkerpas) [inline src/Timer.h:59]: delta? elapsed?

File: src/Timer.h (hunk 26 above)
Comment (rphh) [reply src/Timer.h]: done

File: src/Timer.h (hunk 27 above)
Comment (rphh) [reply src/Timer.h]: changed to elapsedTime

**Review by Lurkerpas — APPROVED** [review]

**Pull request merged**